exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/partitionsender/PartitionerDecorator.java

```diff
@@ -262,68 +280,122 @@ public FlushBatchesHandlingClass(boolean isLastBatch, boolean schemaChanged) {
-      this.latch = latch;
+    public PartitionerTask(PartitionerDecorator partitionerDecorator, GeneralExecuteIface iface, Partitioner partitioner, AtomicInteger count, CountDownLatchInjection testCountDownLatch) {
+      state = new AtomicReference<>(STATE.NEW);
+      runner = new AtomicReference<>();
+      this.partitionerDecorator = partitionerDecorator;
       this.iface = iface;
-      this.part = part;
+      this.partitioner = partitioner;
+      this.count = count;
       this.testCountDownLatch = testCountDownLatch;
     }
 
     @Override
     public void run() {
-      // Test only - Pause until interrupted by fragment thread
-      try {
-        testCountDownLatch.await();
-      } catch (final InterruptedException e) {
-        logger.debug("Test only: partitioner thread is interrupted in test countdown latch await()", e);
-      }
-
-      final Thread currThread = Thread.currentThread();
-      final String currThreadName = currThread.getName();
-      final OperatorStats localStats = part.getStats();
-      try {
-        final String newThreadName = parentThreadName + currThread.getId();
-        currThread.setName(newThreadName);
+      final Thread thread = Thread.currentThread();
+      Preconditions.checkState(runner.compareAndSet(null, thread),
+          "PartitionerTask can be executed only once.");
+      if (state.get() == STATE.NEW) {
+        final String name = thread.getName();
+        thread.setName(String.format("Partitioner-%s-%d", partitionerDecorator.thread.getName(), thread.getId()));
+        final OperatorStats localStats = partitioner.getStats();
         localStats.clear();
         localStats.startProcessing();
-        iface.execute(part);
-      } catch (IOException e) {
-        exp = e;
-      } finally {
-        localStats.stopProcessing();
-        currThread.setName(currThreadName);
-        latch.countDown();
+        ExecutionException executionException = null;
+        try {
+          // Test only - Pause until interrupted by fragment thread
+          testCountDownLatch.await();
+          iface.execute(partitioner);
+        } catch (InterruptedException e) {
+          if (state.compareAndSet(STATE.NEW, STATE.INTERRUPTED)) {
+            logger.warn("Partitioner Task interrupted during the run", e);
+          }
+        } catch (Throwable t) {
+          executionException = new ExecutionException(t);
+        }
+        if (state.compareAndSet(STATE.NEW, STATE.COMPLETING)) {
+          if (executionException == null) {
+            localStats.stopProcessing();
+            state.lazySet(STATE.NORMAL);
+          } else {
+            exception = executionException;
+            state.lazySet(STATE.EXCEPTIONAL);
+          }
+        }
+        if (count.decrementAndGet() == 0) {
+          LockSupport.unpark(partitionerDecorator.thread);
+        }
+        thread.setName(name);
+      }
+      runner.set(null);
+      while (state.get() == STATE.INTERRUPTING) {
```

Comment:
I think we can still leave a dirty interrupt flag. Consider the following case.

 1. We create a PartitionerTask **Task A**.
 2. **Task A** executes and reaches `thread.setName(name)` in the run method.
 3. Now a cancel is called.
 4. `thread.interrupt()` and `state.lazySet(STATE.INTERRUPTED)` is executed.
 5. Now **Task A** can continue executing
 6. **Task A** checks `state.get() == STATE.INTERRUPTING` which is false. So Thread.sleep is never called and the interrupt flag is never cleared.